### PR TITLE
Make ProxyQueryInterface generic

### PR DESCRIPTION
**Pull request by VincentLanglet**
## Subject

This will be for instance useful for batchActions, in order to infer the result of the query.

## Changelog

```markdown
### Added
- Added generics to the ProxyQueryInterface
```

### CODE REVIEW

**Comment by VincentLanglet** [comment]
This would be implemented this way for DoctrineORM
https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/1687

**Comment by VincentLanglet** [comment]
If you have time to take a look @jordisala1991

**Comment by jordisala1991** [comment]
Do we need to change something in the docs like add a phpstan-param to infer the type of execute or is infer automatically?

If is manual, I think it would be nice to add it to docs (in the same example we have now)

**Comment by VincentLanglet** [comment]
> Do we need to change something in the docs like add a phpstan-param to infer the type of execute or is infer automatically?
> 
> If is manual, I think it would be nice to add it to docs (in the same example we have now)

We never add some `phpstan-param` in the doc. Not everyone is using static analysis or even know about it. It might be confusing to add some.

We didn't add any doc about the AdminInterface generic or CrudController. It's just how generics works and people should refer to the phpstan doc instead IMHO.

When you're using Phpstan, on level 6 it will require to add the generic type. So you'll end up having to write
```
@phpstan-param ProxyQueryInterface<Foo>
```
by yourself.

File: tests/Fixtures/Controller/BatchOtherController.php [hunk 1]
```diff
@@ -18,6 +18,9 @@
 
 final class BatchOtherController
 {
+    /**
+     * @param ProxyQueryInterface<object> $aCustomNameForTheProxyQuery
```
Comment (jordisala1991) [inline tests/Fixtures/Controller/BatchOtherController.php:22]: This is what I had in mind, I think it would be nice to be in docs too.

**Review by jordisala1991 — APPROVED** [review]

File: tests/Fixtures/Controller/BatchOtherController.php (hunk 1 above)
Comment (VincentLanglet) [reply tests/Fixtures/Controller/BatchOtherController.php:22]: That's the same idea that when you're doing `Collection<Foo>` or `AdminInterface<Foo>`. 

But there is no documentation in Doctrine, or Symfony or Sonata about generics. I don't think that the place for it.

**Comment by jordisala1991** [comment]
> > Do we need to change something in the docs like add a phpstan-param to infer the type of execute or is infer automatically?
> 
> > 
> 
> > If is manual, I think it would be nice to add it to docs (in the same example we have now)
> 
> 
> 
> We never add some `phpstan-param` in the doc. Not everyone is using static analysis or even know about it. It might be confusing to add some.
> 
> 
> 
> We didn't add any doc about the AdminInterface generic or CrudController. It's just how generics works and people should refer to the phpstan doc instead IMHO.
> 
> 
> 
> When you're using Phpstan, on level 6 it will require to add the generic type. So you'll end up having to write
> 
> ```
> 
> @phpstan-param ProxyQueryInterface<Foo>
> 
> ```
> 
> by yourself.

There could be a few cases:

1. I use phpstan and I like if packages provide me with the examples including the generic thing
2. I dont use phpstan and probably will remove those lines when copying the example code (not a big deal)
3. I dont even know what phpstan is, and I dont even care about phpdoc, so I copy the full example (it wont break his app)
4. I dont copy code, I use it as an example, so I might get useful information about how the code works or how is suposed to be used.

In all cases I think is good to show how are you supose to use Sonata code if you want the best quality. IMO having it on docs, makes people that read docs see that Sonata is improving its quality by providing some nice DX.

I dont say that we add all generics at once in this Pr, but I dont see a reason to not add them, if you use phpstan you will be thankful that we provide complete examples, if you dont you will remove it or not but wont hurt you.

**Pull request merged**